### PR TITLE
Feature/wjiang/add open demo

### PR DESCRIPTION
**Pull request by weiyuan-jiang**
Add a simple program to demonstrate how pFIO open, read and close a file. A user should provide a file named "test_in.nc4" together with the executable.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Trivial change (affects only documentation or cleanup)

## Checklist:

- [x] I have tested this change with a run of GEOSgcm (if non-trivial)
- [x] I have added one of the required labels (0 diff, 0 diff trivial, 0 diff structural, non 0-diff)
- [x] I have updated the CHANGELOG.md accordingly following the style of [Keep a Changelog](https://keepachangelog.com/en/1.0.0/#how)

### CODE REVIEW

**Comment by weiyuan-jiang** [comment]
The error code for "not found file" is (2) but the error code I found online is "-90"

File: pfio/pfio_open_close.F90 [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+  type (FileMetadata) :: file_metadata
+  type (NetCDF4_FileFormatter) :: formatter
+  type (FileMetadata) :: test_metadata
+  integer :: status
+
+  call formatter%open('test_in.nc4', pFIO_READ, rc=status)
+  _ASSERT(status==0, "Wrong in open test_in.nc4")
+  file_metadata = formatter%read(rc=status)
+  _ASSERT(status==0, "Wrong in reading test_in.nc4")
```
Comment (tclune) [inline pfio/pfio_open_close.F90:18]:
```suggestion
  _ASSERT(status==0, "Failed while reading test_in.nc4")
```

File: pfio/pfio_open_close.F90 [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+  type (FileMetadata) :: file_metadata
+  type (NetCDF4_FileFormatter) :: formatter
+  type (FileMetadata) :: test_metadata
+  integer :: status
+
+  call formatter%open('test_in.nc4', pFIO_READ, rc=status)
+  _ASSERT(status==0, "Failed to open 'test_in.nc4'")
+  file_metadata = formatter%read(rc=status)
+  _ASSERT(status==0, "Wrong in reading test_in.nc4")
```
Comment (tclune) [inline pfio/pfio_open_close.F90]:
```suggestion
  _ASSERT(status==0, "Failure while reading 'test_in.nc4'")
```

File: pfio/pfio_open_close.F90 [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  type (FileMetadata) :: test_metadata
+  integer :: status
+
+  call formatter%open('test_in.nc4', pFIO_READ, rc=status)
+  _ASSERT(status==0, "Failed to open 'test_in.nc4'")
+  file_metadata = formatter%read(rc=status)
+  _ASSERT(status==0, "Failed while reading 'test_in.nc4'")
+  call formatter%close(rc=status)
+  _ASSERT(status==0, "Wrong in closing test_in.nc4")
+  print*, "Successfully open, read and close test_in.nc4"
```
Comment (tclune) [inline pfio/pfio_open_close.F90:21]:
```suggestion
  print*, "Successfully opened, read and closed 'test_in.nc4'."
```

**Review by tclune — APPROVED** [review]

**Comment by mathomp4** [comment]
It looks like GNU is failing with:
```
Scanning dependencies of target pfio_collective_demo.x
/root/project/MAPL/pfio/pfio_open_close.F90:1:0:

    1 | #DEFINE I_AM_MAIN
      | 
Error: invalid preprocessing directive #DEFINE
[ 49%] Building Fortran object pfio/CMakeFiles/pfio_server_demo.x.dir/pfio_server_demo.F90.o
[ 50%] Building Fortran object pfio/CMakeFiles/pfio_collective_demo.x.dir/pfio_collective_demo.F90.o
/root/project/MAPL/pfio/pfio_open_close.F90:16:113:

   16 |   _ASSERT(status==0, "Failed to open 'test_in.nc4'")
      |                                                                                                                 1
Error: Symbol 'rc' at (1) has no IMPLICIT type
```

File: pfio/pfio_open_close.F90 [hunk 4]
```diff
@@ -0,0 +1,22 @@
+#DEFINE I_AM_MAIN
```
Comment (mathomp4) [inline pfio/pfio_open_close.F90:1]:
```suggestion
#define I_AM_MAIN
```

**Comment by mathomp4** [comment]
I just committed a change. Let's see!

**Comment by mathomp4** [comment]
Now I broke it on Intel?

**Review by tclune — APPROVED** [review]

**Pull request merged**